util/tracecpu/profile.go

```diff
@@ -0,0 +1,401 @@
+	}
+	collector TopSQLCollector
+}
+
+// NewSQLStatsProfiler create a sqlStatsProfiler.
+func NewSQLStatsProfiler() *sqlStatsProfiler {
+	return &sqlStatsProfiler{
+		taskCh:     make(chan *profileTask, 128),
+		cacheBufCh: make(chan *profileTask, 128),
+	}
+}
+
+func (sp *sqlStatsProfiler) Run() {
+	sp.mu.Lock()
+	defer sp.mu.Unlock()
+	logutil.BgLogger().Info("cpu profiler started")
+	go sp.startCPUProfileWorker()
+	go sp.startAnalyzeProfileWorker()
+}
+
+func (sp *sqlStatsProfiler) SetCollector(c TopSQLCollector) {
+	sp.collector = c
+}
+
+func (sp *sqlStatsProfiler) RegisterSQL(sqlDigest, normalizedSQL string) {
+	if sp.collector == nil {
+		return
+	}
+	sp.collector.RegisterSQL(sqlDigest, normalizedSQL)
+}
+
+func (sp *sqlStatsProfiler) RegisterPlan(planDigest string, normalizedPlan string) {
+	if sp.collector == nil {
+		return
+	}
+	sp.collector.RegisterPlan(planDigest, normalizedPlan)
+}
+
+func (sp *sqlStatsProfiler) startCPUProfileWorker() {
+	defer util.Recover("top-sql", "profileWorker", nil, false)
+	for {
+		if sp.IsEnabled() {
+			sp.doCPUProfile()
+		} else {
+			time.Sleep(time.Second)
+		}
+	}
+}
+
+func (sp *sqlStatsProfiler) doCPUProfile() {
+	interval := config.GetGlobalConfig().TopSQL.RefreshInterval
+	task := sp.newProfileTask()
+	if err := pprof.StartCPUProfile(task.buf); err != nil {
+		// Sleep a while before retry.
+		time.Sleep(time.Millisecond)
+		return
+	}
+	ns := int(time.Second)*interval - time.Now().Nanosecond()
+	time.Sleep(time.Nanosecond * time.Duration(ns))
+	pprof.StopCPUProfile()
+	sp.sendProfileTask(task)
+}
+
+func (sp *sqlStatsProfiler) sendProfileTask(task *profileTask) {
+	task.end = time.Now().Unix()
+	sp.taskCh <- task
+}
+
+func (sp *sqlStatsProfiler) startAnalyzeProfileWorker() {
+	defer util.Recover("top-sql", "analyzeProfileWorker", nil, false)
+	for {
+		task := <-sp.taskCh
+		reader := bytes.NewReader(task.buf.Bytes())
+		p, err := profile.Parse(reader)
+		if err != nil {
+			logutil.BgLogger().Error("parse profile error", zap.Error(err))
+			continue
+		}
+		stats := sp.parseCPUProfileTags(p)
+		sp.handleExportProfileTask(p)
+		if sp.collector != nil {
+			sp.collector.Collect(task.end, stats)
+		}
+		sp.putTaskToBuffer(task)
+	}
+}
+
+type profileTask struct {
+	buf *bytes.Buffer
+	end int64
+}
+
+func (sp *sqlStatsProfiler) newProfileTask() *profileTask {
+	var task *profileTask
+	select {
+	case task = <-sp.cacheBufCh:
+		task.buf.Reset()
+	default:
+		task = &profileTask{
+			buf: bytes.NewBuffer(make([]byte, 0, 100*1024)),
+		}
+	}
+	return task
+}
+
+func (sp *sqlStatsProfiler) putTaskToBuffer(task *profileTask) {
+	select {
+	case sp.cacheBufCh <- task:
+	default:
+	}
+}
+
+func (sp *sqlStatsProfiler) parseCPUProfileTags(p *profile.Profile) []SQLStats {
+	sqlMap := make(map[string]*sqlStats)
+	idx := len(p.SampleType) - 1
+	for _, s := range p.Sample {
+		digests, ok := s.Label[labelSQLDigest]
+		if !ok || len(digests) == 0 {
+			continue
+		}
+		for _, digest := range digests {
```

Comment:
Could add some brief introduction about the Profile sample so we are easier to understand the code?
Maybe something like https://github.com/DataDog/go-profiler-notes/blob/main/pprof.md

Reply:
Already add some comment for this function.